### PR TITLE
Transaction fees & KPIs

### PR DESCRIPTION
**Pull request by ncomerci**
**Do not merge until transaction tags PR is merged.**

- KPIs migrated to code-based computing
- Out transaction fees analysis
- Code refactors

### CODE REVIEW

**Review by 1emu — APPROVED** [review]
Great work Nico 💪 ! 
I left some suggestions and a PR to add some tests.
I'm approving assuming we address those before merging :)

File: src/export-balances.ts [hunk 1]
```diff
@@ -1,53 +1,57 @@
...
-    }))
-    balances.push(...balanceParsed)
+  for (const wallet of Wallets.get()) {
+    unresolvedBalances.push(getBalance(wallet))
   }
 
-  balances = balances.filter(
-    b => ALLOWED_SYMBOLS.includes(b.symbol) && b.amount > 0
-  )
+  let balances = flattenArray(await Promise.all(unresolvedBalances))
```
Comment (andyesp) [inline src/export-balances.ts:52]: You can write this `const balances = flattenArray(...).filter(...)`

File: src/entities/Networks.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  static get(name: `${NetworkName}`): NetworkType {
+    return this.LIST.find(network => network.name === name)
+  }
+
+  static getAll(): NetworkType[] {
+    return this.LIST
+  }
+}
+
+export default Networks
```
Comment (andyesp) [inline src/entities/Networks.ts:33]: I'd change this to an object export to keep the same pattern as the other entities.

File: src/entities/Tags.ts [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+    }
+
+    return secondarySaleTags[this.getItemContract(address)]
+  }
+
+  public static isAPITag(tag: string): boolean {
+    return !!this.TAG_CATEGORIES[tag]
+  }
+
+  public static getAPITagCategory(tag: string): TagCategoryData {
```
[the remark this answers is not in the document]
Comment (andyesp) [reply src/entities/Tags.ts:248]: Or just `getCategory`?

File: src/utils.ts [hunk 4]
```diff
@@ -1,23 +1,48 @@
...
+}
+
+export function avg(array: number[]) {
+  return sum(array) / array.length
+}
+
+export function median(array: number[]) {
+  if (array.length === 0) throw new Error("Median: no inputs")
+
+  array.sort((a, b) => a - b)
```
Comment (andyesp) [inline src/utils.ts:28]: You are modifying the original array here. You can copy it and store it in a new variable.
Maybe doing something like `const sortedArray = [...array].sort(...)`

File: src/export-balances.ts [hunk 5]
```diff
@@ -1,53 +1,57 @@
...
+  const { name, address, network } = wallet
+  const url = `https://api.covalenthq.com/v1/${network.id}/address/${address}/portfolio_v2/?key=${API_KEY}`
+  const json = await fetchURL(url)
+  const contracts: Contract[] = json.data.items
+
+
+  return contracts.map<BalanceParsed>(t => ({
+    timestamp: t.holdings[0].timestamp,
+    name,
+    amount: new BigNumber(t.holdings[0].close.balance).dividedBy(10 ** t.contract_decimals).toNumber(),
```
[the remark this answers is not in the document]
Comment (ncomerci) [reply src/export-balances.ts:35]: There's always a value!

image

File: src/export-kpis.ts [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+
+function getRowsByProposalStatus(proposals: ProposalParsed[]) {
+  return Object.values(Status).map(status => {
+    const count = proposals.filter(p => p.status === status).length
+    return [status, count, getRatio(count, proposals.length)]
+  })
+}
+
+function getVPDistributionRows(members: MemberInfo[], totalVP: number) {
+  const rows: any[] = [['Total Members', members.length, '', Math.round(totalVP)]]
```
[the remark this answers is not in the document]
Comment (ncomerci) [reply src/export-kpis.ts:115]: Yes, it's the place for the members percentage. I don't think it makes much sense to set 100% there

image

**Review by andyesp — APPROVED** [review]

File: src/export-transactions.ts [hunk 7]
```diff
@@ -66,16 +52,16 @@ const OPENSEA_ADDRESSES = new Set([
   '0xf715beb51ec8f63317d66f491e37e7bb048fcc2d'
 ])
 
-async function getTopicTxs(network: number, startblock: number, topic: Topic) {
+async function getTopicTxs(network: Network, startblock: number, topic: Topic) {
   const events: string[] = []
   let block = startblock
-  let url = `https://api.covalenthq.com/v1/${network}/block_v2/latest/?key=${COVALENT_API_KEY}`
+  let url = `https://api.covalenthq.com/v1/${network.id}/block_v2/latest/?key=${COVALENT_API_KEY}`
```
Comment (1emu) [inline src/export-transactions.ts:58]:
```typescript 
function baseCovalentUrl(network:Network){
   return `https://api.covalenthq.com/v1/${network.id}`
}
```

File: src/export-transactions.ts [hunk 8]
```diff
@@ -87,9 +73,8 @@ async function getTopicTxs(network: number, startblock: number, topic: Topic) {
   return events
 }
 
-async function getTransactions(name: string, tokenAddress: string, network: number, address: string) {
-  let token = tokens[tokenAddress]
-  const url = `https://api.covalenthq.com/v1/${network}/address/${address}/transfers_v2/?key=${COVALENT_API_KEY}&contract-address=${tokenAddress}&page-size=500000`
+async function getTransactions(name: string, tokenAddress: string, network: Network, address: string) {
+  const url = `https://api.covalenthq.com/v1/${network.id}/address/${address}/transfers_v2/?key=${COVALENT_API_KEY}&contract-address=${tokenAddress}&page-size=500000`
```
Comment (1emu) [inline src/export-transactions.ts:77]: I know it's a high number, but eventually we are going to need to check that there are no more records in the next page, right?

If that's the case I'm ok with creating the ticket and merging as is.

Comment (1emu) [reply src/export-transactions.ts:77]: maybe use something like Governance's `inBatches`

**Review by 1emu — APPROVED** [review]
🔥

File: src/export-transactions.ts (hunk 8 above)
Comment (ncomerci) [reply src/export-transactions.ts:77]: Glad you mentioned it, I've already included it in my next PR!

image
image

**Pull request merged**